### ISSUE TITLE
cannot build on alpinelinux

### ISSUE THREAD

**Issue by zaoqi** [racket/racket#2797]
```
$ docker run -it --rm alpine
# wget -O - https://mirror.racket-lang.org/installers/7.4/racket-minimal-7.4-src-builtpkgs.tgz | tar -xvz
# apk add gcc libc-dev binutils make
# cd racket-7.4/src
# ./configure
# make -j4
......
gcc -Wall  -I./.. -I./../include  -I./../../rktio -I../../rktio -g -O2  -pthread  -DUSE_SENORA_GC  -c ./syntax.c -o syntax.o
gcc -Wall  -I./.. -I./../include  -I./../../rktio -I../../rktio -g -O2  -pthread  -DUSE_SENORA_GC  -c ./thread.c -o thread.o
gcc -Wall  -I./.. -I./../include  -I./../../rktio -I../../rktio -g -O2  -pthread  -DUSE_SENORA_GC  -c ./type.c -o type.o
gcc -Wall  -I./.. -I./../include  -I./../../rktio -I../../rktio -g -O2  -pthread  -DUSE_SENORA_GC  -c ./unwind/libunwind.c -o unwind.o
In file included from ./unwind/libunwind_i.h:45,
                 from ./unwind/libunwind.c:31:
./unwind/libunwind.h:60: warning: "_XOPEN_SOURCE" redefined
 # define _XOPEN_SOURCE /* needed for Mac OS X */

In file included from /usr/include/inttypes.h:8,
                 from ./unwind/../../mzconfig.h:33,
                 from ./unwind/../../sconfig.h:20,
                 from ./unwind/libunwind.c:26:
/usr/include/features.h:16: note: this is the location of the previous definition
 #define _XOPEN_SOURCE 700

In file included from ./unwind/libunwind.c:31:
./unwind/libunwind_i.h:54: warning: "__BYTE_ORDER" redefined
 # define __BYTE_ORDER __LITTLE_ENDIAN

In file included from /usr/include/sys/types.h:70,
                 from ./unwind/libunwind_i.h:42,
                 from ./unwind/libunwind.c:31:
/usr/include/endian.h:11: note: this is the location of the previous definition
 #define __BYTE_ORDER __BYTE_ORDER__

./unwind/libunwind.c: In function '_Ux86_64_uc_addr':
./unwind/libunwind.c:3085:55: error: 'REG_R8' undeclared (first use in this function); did you mean 'EM_RX'?
     case UNW_X86_64_R8: addr = &uc->uc_mcontext.gregs[REG_R8]; break;
                                                       ^~~~~~
                                                       EM_RX
./unwind/libunwind.c:3085:55: note: each undeclared identifier is reported only once for each function it appears in
./unwind/libunwind.c:3086:55: error: 'REG_R9' undeclared (first use in this function); did you mean 'EM_RX'?
     case UNW_X86_64_R9: addr = &uc->uc_mcontext.gregs[REG_R9]; break;
                                                       ^~~~~~
                                                       EM_RX
./unwind/libunwind.c:3087:56: error: 'REG_R10' undeclared (first use in this function)
     case UNW_X86_64_R10: addr = &uc->uc_mcontext.gregs[REG_R10]; break;
                                                        ^~~~~~~
./unwind/libunwind.c:3088:56: error: 'REG_R11' undeclared (first use in this function)
     case UNW_X86_64_R11: addr = &uc->uc_mcontext.gregs[REG_R11]; break;
                                                        ^~~~~~~
./unwind/libunwind.c:3089:56: error: 'REG_R12' undeclared (first use in this function)
     case UNW_X86_64_R12: addr = &uc->uc_mcontext.gregs[REG_R12]; break;
                                                        ^~~~~~~
./unwind/libunwind.c:3090:56: error: 'REG_R13' undeclared (first use in this function)
     case UNW_X86_64_R13: addr = &uc->uc_mcontext.gregs[REG_R13]; break;
                                                        ^~~~~~~
./unwind/libunwind.c:3091:56: error: 'REG_R14' undeclared (first use in this function)
     case UNW_X86_64_R14: addr = &uc->uc_mcontext.gregs[REG_R14]; break;
                                                        ^~~~~~~
./unwind/libunwind.c:3092:56: error: 'REG_R15' undeclared (first use in this function)
     case UNW_X86_64_R15: addr = &uc->uc_mcontext.gregs[REG_R15]; break;
                                                        ^~~~~~~
./unwind/libunwind.c:3093:56: error: 'REG_RDI' undeclared (first use in this function)
     case UNW_X86_64_RDI: addr = &uc->uc_mcontext.gregs[REG_RDI]; break;
                                                        ^~~~~~~
./unwind/libunwind.c:3094:56: error: 'REG_RSI' undeclared (first use in this function)
     case UNW_X86_64_RSI: addr = &uc->uc_mcontext.gregs[REG_RSI]; break;
                                                        ^~~~~~~
./unwind/libunwind.c:3095:56: error: 'REG_RBP' undeclared (first use in this function)
     case UNW_X86_64_RBP: addr = &uc->uc_mcontext.gregs[REG_RBP]; break;
                                                        ^~~~~~~
./unwind/libunwind.c:3096:56: error: 'REG_RBX' undeclared (first use in this function); did you mean 'EM_RX'?
     case UNW_X86_64_RBX: addr = &uc->uc_mcontext.gregs[REG_RBX]; break;
                                                        ^~~~~~~
                                                        EM_RX
./unwind/libunwind.c:3097:56: error: 'REG_RDX' undeclared (first use in this function); did you mean 'EM_RX'?
     case UNW_X86_64_RDX: addr = &uc->uc_mcontext.gregs[REG_RDX]; break;
                                                        ^~~~~~~
                                                        EM_RX
./unwind/libunwind.c:3098:56: error: 'REG_RAX' undeclared (first use in this function); did you mean 'EM_RX'?
     case UNW_X86_64_RAX: addr = &uc->uc_mcontext.gregs[REG_RAX]; break;
                                                        ^~~~~~~
                                                        EM_RX
./unwind/libunwind.c:3099:56: error: 'REG_RCX' undeclared (first use in this function); did you mean 'EM_RCE'?
     case UNW_X86_64_RCX: addr = &uc->uc_mcontext.gregs[REG_RCX]; break;
                                                        ^~~~~~~
                                                        EM_RCE
./unwind/libunwind.c:3100:56: error: 'REG_RSP' undeclared (first use in this function); did you mean 'EM_ZSP'?
     case UNW_X86_64_RSP: addr = &uc->uc_mcontext.gregs[REG_RSP]; break;
                                                        ^~~~~~~
                                                        EM_ZSP
./unwind/libunwind.c:3101:56: error: 'REG_RIP' undeclared (first use in this function)
     case UNW_X86_64_RIP: addr = &uc->uc_mcontext.gregs[REG_RIP]; break;
                                                        ^~~~~~~
make[7]: *** [Makefile:295: unwind.o] Error 1
make[7]: *** Waiting for unfinished jobs....
make[7]: Leaving directory '/racket-7.4/src/racket/src'
make[6]: *** [Makefile:138: all] Error 2
make[6]: Leaving directory '/racket-7.4/src/racket/src'
make[5]: *** [Makefile:159: mzlibrary] Error 2
make[5]: Leaving directory '/racket-7.4/src/racket'
make[4]: *** [Makefile:99: cgc-core] Error 2
make[4]: Leaving directory '/racket-7.4/src/racket'
make[3]: *** [Makefile:93: cgc] Error 2
make[3]: Leaving directory '/racket-7.4/src/racket'
make[2]: *** [Makefile:107: 3m] Error 2
make[2]: Leaving directory '/racket-7.4/src/racket'
make[1]: *** [Makefile:61: 3m] Error 2
make[1]: Leaving directory '/racket-7.4/src'
make: *** [Makefile:51: all] Error 2
```

**Comment by pmatos** [comment]
I can reproduce this.
@zaoqi thanks for the issue

**Comment by pmatos** [comment]
The next issue seems to be `__BYTE_ORDER`:

```
In file included from ./unwind/libunwind.c:31:0:
./unwind/libunwind_i.h:54:0: warning: "__BYTE_ORDER" redefined
 # define __BYTE_ORDER __LITTLE_ENDIAN
 
In file included from /usr/include/sys/types.h:70:0,
                 from ./unwind/libunwind_i.h:42,
                 from ./unwind/libunwind.c:31:
/usr/include/endian.h:11:0: note: this is the location of the previous definition
 #define __BYTE_ORDER __BYTE_ORDER__
 
./unwind/libunwind.c: In function '_Ux86_64_uc_addr':
./unwind/libunwind.c:3085:55: error: 'REG_R8' undeclared (first use in this function)
     case UNW_X86_64_R8: addr = &uc->uc_mcontext.gregs[REG_R8]; break;
...
```

**Comment by pmatos** [comment]
Once #2811 is merged, things get tricky. 
Build fails with missing `getcontext`. glibc defined `getcontext` and friends but musl doesn't.

For good reason since they have been deprecated so it seems we might never have these in musl. This is a problem also because sooner or later they might disappear from glibc. These are currently used by our slice of libunwind.

**Comment by samth** [comment]
See https://github.com/libunwind/libunwind/issues/33 and linked issues for more on this, suggesting that maybe it can work on x86_64.

**Comment by pmatos** [comment]
@samth yes, I read that one in my research into the issue. Current libunwind (as mention in the issue) has an impl. of `getcontext` and friends for this specific case. However, our slice doesn't have it. I am wondering how hard it would be to ditch our slice and use the latest libunwind. Carrying this kind of baggage in the racket source feel undesirable precisely for this reason even though I can imagine there were good reasons to initially include a slice in our sources. Maybe @mflatt can also chime in on this.

**Comment by pmatos** [comment]
libunwind slice was imported here: https://github.com/samth/plt-history/commit/961d459a7d6f156119f6f6b45c7e6dc5a6edab6c

There haven't been a significant amount of changes to our slice afaict, but libunwind upstream moved significantly. Would the team be happy with a PR attempting to replace the slice by the use of upstream libunwind? @samth @mflatt? Do you see any blockers to doing this?

**Comment by mflatt** [comment]
Does modern libunwind provide a way to deal with frames for JIT-generated code, which doesn't necessarily follow C conventions and that certainly doesn't have DWARF information? And can it always avoid dereferencing a bad pointer (e.g., when the stack is truncated because it's a partial copy due to a continuation application)?  I think those were probably the issues that triggered the slice/fork before, but I forget the details.

It make may more sense to just disable libunwind-based stack traces on platforms where it doesn't work right, and look to the future through Racket CS.

**Comment by pmatos** [comment]
> Does modern libunwind provide a way to deal with frames for JIT-generated code, which doesn't necessarily follow C conventions and that certainly doesn't have DWARF information? And can it always avoid dereferencing a bad pointer (e.g., when the stack is truncated because it's a partial copy due to a continuation application)? I think those were probably the issues that triggered the slice/fork before, but I forget the details.

Thanks for the quick rundown of issues. This is something that can be investigated.

> 
> It make may more sense to just disable libunwind-based stack traces on platforms where it doesn't work right, and look to the future through Racket CS.

However, that makes sense. If we are going to look into the future in RacketCS, then  we should spend our limited time working on improving RacketCS instead.

**Comment by pmatos** [comment]
This has been fixed at 843824c94e2c5d9518e30a54c32b1dcf81154a7b
Now building in `alpine:latest` with musl.
